circleci: failed tests: TestPurgeOldLeases

The following test appears to have failed:

[#18127](https://circleci.com/gh/cockroachdb/cockroach/18127):

```
E160519 20:25:59.145658 storage/queue.go:446  [split] range=6 [/Table/50-/Max): error: storage/split_queue.go:103: unable to split range=6 [/Table/50-/Max) at key "/Table/51/0": storage/replica_command.go:2064: split at key /Table/51 failed: node unavailable; try another peer
I160519 20:26:00.143159 /go/src/google.golang.org/grpc/clientconn.go:499  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:37408: getsockopt: connection refused"; Reconnecting to "127.0.0.1:37408"
I160519 20:26:01.567588 /go/src/google.golang.org/grpc/clientconn.go:499  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:37408: getsockopt: connection refused"; Reconnecting to "127.0.0.1:37408"
I160519 20:26:01.568134 /go/src/google.golang.org/grpc/clientconn.go:351  Failed to dial 127.0.0.1:37408: grpc: timed out trying to connect; please retry.
I160519 20:26:01.568497 storage/engine/rocksdb.go:178  closing in-memory rocksdb instance
--- FAIL: TestPurgeOldLeases (3.06s)
	lease_internal_test.go:148: found 2 leases instead of 3
=== RUN   TestNameCacheIsUpdated
W160519 20:26:01.626028 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160519 20:26:01.632256 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160519 20:26:01.636174 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160519 20:26:01.637370 server/node.go:401  store store=0:0 ([]=) not bootstrapped
I160519 20:26:01.645034 storage/replica_command.go:1470  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406580h26m2.643877175s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-19 20:26:01.644854963 +0000 UTC]
I160519 20:26:01.645633 storage/replica.go:1524  reproposing 1 commands after empty entry
E160519 20:26:01.647609 storage/replica.go:1615  [node=1,store=1,range=1] error executing raft command: write at timestamp 1463689561.643683939,0 too old; wrote at 1463689561.643683939,1
I160519 20:26:01.648572 server/node.go:327  **** cluster {1ce91edf-ae90-4afe-bbaf-8ec260ecfe6e} has been created
--
I160519 20:29:01.169010 kv/dist_sender.go:974  application error: retry txn "sql/executor.go:420 sql txn" id=228a2eaa key=/Table/51/1/"g"/0 rw=true pri=0.00407542 iso=SERIALIZABLE stat=PENDING epo=0 ts=1463689741.081965649,0 orig=1463689741.081965649,0 max=1463689741.081965649,0 wto=false rop=false
W160519 20:29:01.169171 kv/dist_sender.go:719  failed to invoke EndTransaction [/Table/51/1/"g"/0,/Min): retry txn "sql/executor.go:420 sql txn" id=228a2eaa key=/Table/51/1/"g"/0 rw=true pri=0.00407542 iso=SERIALIZABLE stat=PENDING epo=0 ts=1463689741.081965649,0 orig=1463689741.081965649,0 max=1463689741.081965649,0 wto=false rop=false
I160519 20:29:01.169579 client/db.go:474  failed batch: retry txn "sql/executor.go:420 sql txn" id=228a2eaa key=/Table/51/1/"g"/0 rw=true pri=0.01335596 iso=SERIALIZABLE stat=PENDING epo=1 ts=1463689741.081965649,0 orig=1463689741.081965649,0 max=1463689741.131965649,0 wto=false rop=false
I160519 20:29:01.175687 storage/engine/rocksdb.go:178  closing in-memory rocksdb instance
--- PASS: TestRollbackInRestartWait (1.13s)
FAIL
FAIL	github.com/cockroachdb/cockroach/sql	183.317s
I160519 20:26:12.790986 rand.go:76  Random seed: 5117411435364211007
=== RUN   TestProcessExpression
--- PASS: TestProcessExpression (0.01s)
=== RUN   TestServer
W160519 20:26:12.834804 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160519 20:26:12.862682 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160519 20:26:12.869207 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160519 20:26:12.870734 server/node.go:401  store store=0:0 ([]=) not bootstrapped
I160519 20:26:12.889583 storage/replica_command.go:1470  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406580h26m13.888032954s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-19 20:26:12.889276071 +0000 UTC]
I160519 20:26:12.892139 server/node.go:327  **** cluster {3eba6378-dc1b-4650-a26f-a213ef53c4df} has been created

```
Please assign, take a look and update the issue accordingly.

--- a/sql/lease.go
+++ b/sql/lease.go
@@ -747,6 +747,10 @@ func (t *tableState) purgeOldLeases(
 
 // LeaseManagerTestingKnobs contains test affordances.
 type LeaseManagerTestingKnobs struct {
+	// A callback called when a gossip update is received, before the leases are
+	// refreshed. Careful when using this to block for too long - you can block
+	// all the gossip users in the system.
+	GossipUpdateEvent func(config.SystemConfig)
 	// A callback called after the leases are refreshed as a result of a gossip update.
 	TestingLeasesRefreshedEvent func(config.SystemConfig)
 }
@@ -1065,6 +1069,9 @@ func (m *LeaseManager) RefreshLeases(s *stop.Stopper, db *client.DB, gossip *gos
 			select {
 			case <-gossipUpdateC:
 				cfg, _ := gossip.GetSystemConfig()
+				if m.testingKnobs.GossipUpdateEvent != nil {
+					m.testingKnobs.GossipUpdateEvent(cfg)
+				}
 				// Read all tables and their versions
 				if log.V(2) {
 					log.Info("received a new config; will refresh leases")

--- a/sql/lease_internal_test.go
+++ b/sql/lease_internal_test.go
@@ -23,7 +23,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/cockroach/base"
 	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/config"
+	"github.com/cockroachdb/cockroach/server/testingshim"
 	"github.com/cockroachdb/cockroach/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/util/leaktest"
@@ -112,11 +115,36 @@ func TestLeaseSet(t *testing.T) {
 	}
 }
 
+func getNumLeases(ts *tableState) int {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	return len(ts.active.data)
+}
+
 func TestPurgeOldLeases(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	s, db, kvDB, cleanup := sqlutils.SetupServer(t)
+	// We're going to block gossip so it doesn't come randomly and clear up the
+	// leases we're artificially setting up.
+	gossipSem := make(chan struct{}, 1)
+	serverParams := testingshim.TestServerParams{
+		Knobs: base.TestingKnobs{
+			SQLLeaseManager: &LeaseManagerTestingKnobs{
+				GossipUpdateEvent: func(cfg config.SystemConfig) {
+					gossipSem <- struct{}{}
+					<-gossipSem
+				},
+			},
+		},
+	}
+	s, db, kvDB, cleanup := sqlutils.SetupServerWithParams(t, serverParams)
 	defer cleanup()
 	leaseManager := s.LeaseManager().(*LeaseManager)
+	// Block gossip.
+	gossipSem <- struct{}{}
+	defer func() {
+		// Unblock gossip.
+		<-gossipSem
+	}()
 
 	if _, err := db.Exec(`
 CREATE DATABASE t;
@@ -124,6 +152,7 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 `); err != nil {
 		t.Fatal(err)
 	}
+
 	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
 
 	var leases []*LeaseState
@@ -144,7 +173,7 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 		t.Fatal(err)
 	}
 	ts := leaseManager.findTableState(tableDesc.ID, false, nil)
-	if numLeases := len(ts.active.data); numLeases != 3 {
+	if numLeases := getNumLeases(ts); numLeases != 3 {
 		t.Fatalf("found %d leases instead of 3", numLeases)
 	}
 
@@ -153,10 +182,13 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 		t.Fatal(err)
 	}
 
-	if numLeases := len(ts.active.data); numLeases != 1 {
+	if numLeases := getNumLeases(ts); numLeases != 1 {
 		t.Fatalf("found %d leases instead of 1", numLeases)
 	}
-	if ts.active.data[0] != leases[2] {
+	ts.mu.Lock()
+	correctLease := ts.active.data[0] == leases[2]
+	ts.mu.Unlock()
+	if !correctLease {
 		t.Fatalf("wrong lease survived purge")
 	}
 }

